mailbox/backup/src/main/java/org/apache/james/mailbox/backup/DefaultMailboxBackup.java

```diff
@@ -81,8 +101,112 @@ public void backupAccount(User user, OutputStream destination) throws IOExceptio
+                break;
+            case MAILBOX:
+                MailboxWithAnnotationsArchiveEntry mailboxArchiveEntry = ((MailboxWithAnnotationsArchiveEntry) entry);
+                LOGGER.error(getMailboxEntryInWrongPositionErrorMessage(mailboxArchiveEntry));
+                break;
+            case UNKNOWN:
+                String entryName = ((UnknownArchiveEntry) entry).getEntryName();
+                LOGGER.error("unknow entry found in zip :" + entryName);
+                break;
+        }
+    }
+
+    private String getMailboxEntryInWrongPositionErrorMessage(MailboxWithAnnotationsArchiveEntry mailboxArchiveEntry) {
+        return "mailbox entry found in wrong position in zip, should be before messages." + " id :"
+            + mailboxArchiveEntry.getMailboxId().getValue()
+            + " - name : " + mailboxArchiveEntry.getMailboxName();
+    }
+
+    private void restoreMessageEntryWithErrorHandling(MailboxSession session, Map<SerializedMailboxId, MessageManager> oldToNewMailbox, MessageArchiveEntry messageArchiveEntry) {
+        try {
+            restoreMessageEntry(session, oldToNewMailbox, messageArchiveEntry);
+        } catch (IOException ioException) {
+            LOGGER.error("IO error when restoring message :" + messageArchiveEntry.getMessageId());
+        } catch (MailboxException mailboxException) {
+            LOGGER.error("Mailbox error when restoring message :" + messageArchiveEntry.getMessageId());
+        }
+    }
+
+    private void restoreMessageEntry(MailboxSession session, Map<SerializedMailboxId, MessageManager> oldToNewMailbox,
+                                     MessageArchiveEntry messageArchiveEntry) throws IOException, MailboxException {
+        try (InputStream content = messageArchiveEntry.getContent()) {
+            MessageManager mailbox = oldToNewMailbox.get(messageArchiveEntry.getMailboxId());
+            MessageManager.AppendCommand appendCommand = MessageManager.AppendCommand.builder()
+                .withFlags(messageArchiveEntry.getFlags())
+                .withInternalDate(messageArchiveEntry.getInternalDate())
+                .build(content);
+            mailbox.appendMessage(appendCommand, session);
+        }
+    }
+
+    private Map<SerializedMailboxId, MessageManager> restoreMailboxes(MailboxSession session,
+                                                                      ImmutablePair<List<MailboxWithAnnotationsArchiveEntry>, Optional<MessageArchiveEntry>> mailboxesAndFirstMessage) {
+        return mailboxesAndFirstMessage.left.stream()
+            .flatMap(Throwing.function(mailboxEntry ->
+                OptionalUtils.toStream(restoreMailboxEntry(session, mailboxEntry))))
+            .collect(Guavate.entriesToImmutableMap());
+    }
+
+    private ImmutablePair<List<MailboxWithAnnotationsArchiveEntry>, Optional<MessageArchiveEntry>> readMailboxes(MailArchiveIterator iterator) {
+        List<MailboxWithAnnotationsArchiveEntry> mailboxes = new ArrayList<>();
```

Comment:
We can use ImmutableList.Builder here instead of ArrayList. Then just need to .build() at returning and we don't have to use ImmutableList.copyOf() - which I think it would consum more memory